### PR TITLE
Convert sample to kotlin 

### PR DESCRIPTION
**Pull request by brianPlummer**
this converts the sample to kotlin and switches over the data model to kotlin data classes using moshi's codegen.  switching to moshi for the middleware parsing.

### CODE REVIEW

**Comment by brianPlummer** [comment]
@digitalbuddha

File: app/src/main/java/com/nytimes/android/sample/SampleApp.kt [hunk 1]
```diff
@@ -2,34 +2,30 @@ package com.nytimes.android.sample
...
 import java.util.concurrent.TimeUnit
 
 class SampleApp : Application() {
 
     var nonPersistedStore: Store<RedditData, BarCode>? = null
-    var  persistedStore: Store<RedditData, BarCode>? =null
-    private var persister: Persister<BufferedSource, BarCode>? =null
+    var  persistedStore: Store<RedditData, BarCode>? = null
+    val moshi = Moshi.Builder().build()
+    private var persister: Persister<BufferedSource, BarCode>? = null
```
Comment (ramonaharrison) [inline app/src/main/java/com/nytimes/android/sample/SampleApp.kt]: Could we make all of these Stores and Persisters `lateinit` and non-nullable?

File: app/src/main/java/com/nytimes/android/sample/PersistingStoreActivity.kt [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.schedulers.Schedulers
+import kotlinx.android.synthetic.main.activity_store.*
+
+
+class PersistingStoreActivity : AppCompatActivity() {
+
+    private var postAdapter: PostAdapter? = null
+    private var persistedStore: Store<RedditData, BarCode>? = null
+    private var moshi: Moshi? = null
```
Comment (ramonaharrison) [inline app/src/main/java/com/nytimes/android/sample/PersistingStoreActivity.kt]: Could we make these `lateinit` and non-nullable?

File: app/src/main/java/com/nytimes/android/sample/PersistingStoreActivity.kt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+                .toList()
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe({ showPosts(it) }, {it -> Log.e(PersistingStoreActivity::class.java.simpleName, it.message, it)})
+    }
+
+    private fun showPosts(posts: List<Post>) {
+        postAdapter!!.setPosts(posts)
+        makeText(this@PersistingStoreActivity,
+                "Loaded " + posts.size + " posts",
```
Comment (ramonaharrison) [inline app/src/main/java/com/nytimes/android/sample/PersistingStoreActivity.kt]: just a nitpick: could be cleaned up a bit to use an interpolated string.

File: app/src/main/java/com/nytimes/android/sample/reddit/PostViewHolder.kt [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+class PostViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+
+    fun onBind(article: Post) {
+        itemView.title!!.text = article.title
+        article.nestedThumbnail()?.let { showImage(article) }
+    }
+
+    private fun showImage(article: Post) {
+        val nestedImage = article.nestedThumbnail()
+        nestedImage?.url?.let {
```
Comment (ramonaharrison) [inline app/src/main/java/com/nytimes/android/sample/reddit/PostViewHolder.kt]: why not just use the `let` in `onBind` to pass the non-null `url` in as a parameter to this function?

**Review by ramonaharrison — COMMENTED** [review]
Thank you so much @brianPlummer for starting this enormous undertaking!!!

We'll get a few more eyes from our team on these PRs shortly...

**Review by goldstei — APPROVED** [review]

File: app/src/main/java/com/nytimes/android/sample/PersistingStoreActivity.kt (hunk 2 above)
Comment (brianPlummer) [reply app/src/main/java/com/nytimes/android/sample/PersistingStoreActivity.kt]: done!

File: app/src/main/java/com/nytimes/android/sample/PersistingStoreActivity.kt (hunk 3 above)
Comment (brianPlummer) [reply app/src/main/java/com/nytimes/android/sample/PersistingStoreActivity.kt]: done!

File: app/src/main/java/com/nytimes/android/sample/reddit/PostViewHolder.kt (hunk 4 above)
Comment (brianPlummer) [reply app/src/main/java/com/nytimes/android/sample/reddit/PostViewHolder.kt]: done!

**Pull request merged**